Missing appium-doctor in node_modules

Someone reported an issue on the main appium repo, I discovered the problem and was directed to open an issue here instead. For reference: https://github.com/appium/appium/issues/6466

Copying my comment on the issue:

------

When clicking the doctor button in the app, then issuing appium-doctor in that same terminal, this is the output:

--------
Last login: Wed May  4 14:29:34 on ttys003
gs-mbp15-dward:~ dward$ '/Applications/Appium.app/Contents/Resources/node/bin/node' '/Applications/Appium.app/Contents/Resources/node_modules/appium/bin/appium-doctor.js'
module.js:341
    throw err;
    ^

Error: Cannot find module '/Applications/Appium.app/Contents/Resources/node_modules/appium/bin/appium-doctor.js'
    at Function.Module._resolveFilename (module.js:339:15)
    at Function.Module._load (module.js:290:25)
    at Function.Module.runMain (module.js:447:10)
    at startup (node.js:146:18)
    at node.js:404:3
gs-mbp15-dward:~ dward$ appium-doctor
info AppiumDoctor ### Diagnostic starting ###
info AppiumDoctor  ✔ Xcode is installed at: /Applications/Xcode.app/Contents/Developer
info AppiumDoctor  ✔ Xcode Command Line Tools are installed.
info AppiumDoctor  ✔ DevToolsSecurity is enabled.
info AppiumDoctor  ✔ The Authorization DB is set up properly.
info AppiumDoctor  ✔ The Node.js binary was found at: /usr/local/bin/node
info AppiumDoctor  ✔ HOME is set to: /Users/dward
info AppiumDoctor  ✔ ANDROID_HOME is set to: /Users/dward/Library/Android/sdk
info AppiumDoctor  ✔ JAVA_HOME is set to: /Library/Java/JavaVirtualMachines/jdk1.7.0_79.jdk/Contents/Home
info AppiumDoctor  ✔ adb exists at: /Users/dward/Library/Android/sdk/platform-tools/adb
info AppiumDoctor  ✔ android exists at: /Users/dward/Library/Android/sdk/tools/android
info AppiumDoctor  ✔ emulator exists at: /Users/dward/Library/Android/sdk/tools/emulator
info AppiumDoctor ### Diagnostic completed, no fix needed. ###
info AppiumDoctor 
info AppiumDoctor Everything looks good, bye!
info AppiumDoctor 
gs-mbp15-dward:~ dward$

------

Upon looking in the app itself, it looks as if appium-doctor wasn't packaged with the newest app and the old one it was.

The only file in /Applications/Appium.app/Contents/Resources/node_modules/appium/bin/ is ios-webkit-debug-proxy-launcher.js.

I launched the Appium : 1.5.2,  have all set up done, and which I click the dr. icon,   get this issue  "'/usr/local/bin/node' '/Applications/Appium.app/Contents/Resources/node_modules/appium/bin/appium-doctor.js'
module.js:440
    throw err;
    ^

Error: Cannot find module '/Applications/Appium.app/Contents/Resources/node_modules/appium/bin/appium-doctor.js'
    at Function.Module._resolveFilename (module.js:438:15)
    at Function.Module._load (module.js:386:25)
    at Function.Module.runMain (module.js:575:10)
    at startup (node.js:159:18)
    at node.js:444:3
"

I am also facing the same problem after i upgrade to 1.5.2. Please find the error below

'/Applications/Appium.app/Contents/Resources/node/bin/node' '/Applications/Appium.app/Contents/Resources/node_modules/appium/bin/appium-doctor.js'
module.js:341
    throw err;
    ^

Error: Cannot find module '/Applications/Appium.app/Contents/Resources/node_modules/appium/bin/appium-doctor.js'
    at Function.Module._resolveFilename (module.js:339:15)
    at Function.Module._load (module.js:290:25)
    at Function.Module.runMain (module.js:447:10)
    at startup (node.js:146:18)
    at node.js:404:3

--- a/Appium/AppiumAppDelegate.m
+++ b/Appium/AppiumAppDelegate.m
@@ -95,10 +95,11 @@
 {
     // check is nodejs, appium, or appium pre-reqs are missing
     NSString *nodeRootPath = [[NSBundle mainBundle] resourcePath];
-    BOOL installationRequired = ![NodeInstance instanceExistsAtPath:nodeRootPath];
-    installationRequired |= ![NodeInstance packageIsInstalledAtPath:nodeRootPath withName:@"appium"];
-
-    if (installationRequired)
+    BOOL nodeInstallationRequired = ![NodeInstance instanceExistsAtPath:nodeRootPath];
+    BOOL appiumInstallationRequired = ![NodeInstance packageIsInstalledAtPath:nodeRootPath withName:@"appium"];
+	BOOL doctorInstallationRequired = ![NodeInstance packageIsInstalledAtPath:nodeRootPath withName:@"appium-doctor"];
+	
+    if (nodeInstallationRequired || appiumInstallationRequired || doctorInstallationRequired)
     {
         // install software
         AppiumInstallationWindowController *installationWindow = [[AppiumInstallationWindowController alloc] initWithWindowNibName:@"AppiumInstallationWindow"];
@@ -107,9 +108,19 @@
         [[installationWindow messageLabel] performSelectorOnMainThread:@selector(setStringValue:) withObject:@"Installing NodeJS..." waitUntilDone:YES];
         [[self mainWindowController] setNode:[[NodeInstance alloc] initWithPath:nodeRootPath]];
         [[installationWindow messageLabel] performSelectorOnMainThread:@selector(setStringValue:) withObject:@"Installing Appium..." waitUntilDone:YES];
-        NSString *appiumVersion = [[[[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"] componentsSeparatedByString:@" "] objectAtIndex:0];
-		[[[self mainWindowController] node] installPackage:@"appium" atVersion:appiumVersion forceInstall:NO];
-        [[installationWindow window] performSelectorOnMainThread:@selector(close) withObject:nil waitUntilDone:YES];
+		NSString *appiumVersion = [[[[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"] componentsSeparatedByString:@" "] objectAtIndex:0];
+		
+		if (appiumInstallationRequired)
+		{
+			[[[self mainWindowController] node] installPackage:@"appium" atVersion:appiumVersion forceInstall:NO];
+		}
+		
+		if (doctorInstallationRequired)
+		{
+			[[[self mainWindowController] node] installPackage:@"appium-doctor" atVersion:nil forceInstall:NO];
+		}
+		
+		[[installationWindow window] performSelectorOnMainThread:@selector(close) withObject:nil waitUntilDone:YES];
     }
     else
     {

--- a/Appium/Misc/Utility/NodeInstance.m
+++ b/Appium/Misc/Utility/NodeInstance.m
@@ -62,25 +62,13 @@
 
 	NSString *nodeModulesDirectory = [NSString stringWithFormat:@"%@/%@", _nodeRootPath,@"node_modules"];
     NSString *packagePath = [NSString stringWithFormat:@"%@/%@", nodeModulesDirectory, packageName];
-    if (forceInstall || ![[NSFileManager defaultManager] fileExistsAtPath:packagePath])
-    {
-		// create node modules folder if it does not exist (so npm will install locally)
-		BOOL isDirectory;
-		if(![[NSFileManager defaultManager] fileExistsAtPath:nodeModulesDirectory isDirectory:&isDirectory])
-		{
-			if(![[NSFileManager defaultManager] createDirectoryAtPath:nodeModulesDirectory withIntermediateDirectories:YES attributes:nil error:NULL])
-			{
-				NSLog(@"Error: Create folder failed %@", nodeModulesDirectory);
-			}
-			else
-			{
-				// install package
-				NSString *npmPath = [NSString stringWithFormat:@"%@/%@", _nodeRootPath, @"node/bin/npm"];
-				NSString *packageArg = [NSString stringWithFormat:@"%@%@", packageName, version ? [NSString stringWithFormat:@"%@%@", @"@", version] : @""];
-				[Utility runTaskWithBinary:npmPath arguments:[NSArray arrayWithObjects: @"install", packageArg, nil] path:_nodeRootPath];
-			}
-		}
-    }
+	if (forceInstall || ![[NSFileManager defaultManager] fileExistsAtPath:packagePath])
+	{
+		// install package
+		NSString *npmPath = [NSString stringWithFormat:@"%@/%@", _nodeRootPath, @"node/bin/npm"];
+		NSString *packageArg = [NSString stringWithFormat:@"%@%@", packageName, version ? [NSString stringWithFormat:@"%@%@", @"@", version] : @""];
+		[Utility runTaskWithBinary:npmPath arguments:[NSArray arrayWithObjects: @"install", packageArg, nil] path:_nodeRootPath];
+	}
 }
 
 -(NSString*) pathtoPackage:(NSString*)packageName

--- a/Appium/Model/AppiumModel.m
+++ b/Appium/Model/AppiumModel.m
@@ -139,11 +139,11 @@ BOOL _isServerListening;
 	
 	if (self.developer.developerMode && self.developer.useExternalNodeJSBinary)
 	{
-		command = [NSMutableString stringWithFormat:@"'%@'%@ build/lib/main.js", self.developer.externalNodeJSBinaryPath, nodeDebugArguments];
+		command = [NSMutableString stringWithFormat:@"'%@'%@ appium/build/lib/main.js", self.developer.externalNodeJSBinaryPath, nodeDebugArguments];
 	}
 	else
 	{
-		command = [NSMutableString stringWithFormat:@"'%@%@'%@ build/lib/main.js", [[NSBundle mainBundle] resourcePath], @"/node/bin/node", nodeDebugArguments];
+		command = [NSMutableString stringWithFormat:@"'%@%@'%@ appium/build/lib/main.js", [[NSBundle mainBundle] resourcePath], @"/node/bin/node", nodeDebugArguments];
 	}
 	
 #pragma mark General Preferences
@@ -613,11 +613,11 @@ BOOL _isServerListening;
 	NSString *nodeCommandString;
 	if (self.developer.useExternalNodeJSBinary)
 	{
-		nodeCommandString = [NSString stringWithFormat:@"'%@' bin/appium-doctor.js --port 4722", self.developer.externalNodeJSBinaryPath];
+		nodeCommandString = [NSString stringWithFormat:@"'%@' appium-doctor/appium-doctor.js --port 4722", self.developer.externalNodeJSBinaryPath];
 	}
 	else
 	{
-		nodeCommandString = [NSString stringWithFormat:@"'%@%@' bin/appium-doctor.js --port 4722", [[NSBundle mainBundle]resourcePath], @"/node/bin/node"];
+		nodeCommandString = [NSString stringWithFormat:@"'%@%@' appium-doctor/appium-doctor.js --port 4722", [[NSBundle mainBundle]resourcePath], @"/node/bin/node"];
 		
 	}
     
@@ -634,7 +634,7 @@ BOOL _isServerListening;
 
 - (void)startExternalDoctor
 {
-	NSString *doctorPath = [[[NSBundle mainBundle] resourcePath] stringByAppendingString:@"/node_modules/appium/bin/appium-doctor.js"];
+	NSString *doctorPath = [[[NSBundle mainBundle] resourcePath] stringByAppendingString:@"/node_modules/appium-doctor/appium-doctor.js"];
     
 	NSString *command;
     
@@ -661,7 +661,7 @@ BOOL _isServerListening;
 	}
 	else
 	{
-		[self.serverTask setCurrentDirectoryPath:[NSString stringWithFormat:@"%@/%@", [[NSBundle mainBundle]resourcePath], @"node_modules/appium"]];
+		[self.serverTask setCurrentDirectoryPath:[NSString stringWithFormat:@"%@/%@", [[NSBundle mainBundle]resourcePath], @"node_modules"]];
 	}
 	
 	// Add a cd call to the start of the command in case the .bash_profile or .bashrc changes the current directory

